### PR TITLE
Update (some) breaking API changes

### PR DESCRIPTION
**Pull request by BenjaminRodenberg**
I updated some of the breaking API changes here. Mainly

* Replacing old read/writeData functions with new ones.
* Replacing raw pointers with spans
* Replacing SolverInterface with Participant
* Removal of outdated API functions

I also added some todos where there are still things to be done. I will merge this PR (independent from the review) soon, because otherwise working on other PRs becomes a real pain (e.g. #257). Since `precice-v3` is a non-released develop branch and the current state is incomplete and inconsistent anyway, I think there is no real harm.

If anybody sees potential for improvement: Please directly apply the changes here. There is no need for long discussions in my opinion.

### CODE REVIEW

File: pages/docs/configuration/configuration-mapping.md [hunk 1]
```diff
@@ -96,7 +96,7 @@ For a complete overview of all basis function, refer to [this paper](https://www
 The interpolation problem might not be well-defined if you map along an axis-symmetric surface. This means, preCICE tries to compute, for example, a 3D interpolant out of 2D information. If so, preCICE throws an error `RBF linear system has not converged` or `Interpolation matrix C is not invertible`. In this case, you can restrict the interpolation problem by ignoring certain coordinates, e.g. `x-dead="true"` to ignore the x coordinate.
 
 {% note %}
-All data mappings are executed during `advance` and not in `readBlockVectorData` etc., cf. the section on  [how to couple your own code](couple-your-code-overview.html).
+All data mappings are executed during `advance` and not in `readData` etc., cf. the section on  [how to couple your own code](couple-your-code-overview.html).
```
Comment (MakisH) [inline pages/docs/configuration/configuration-mapping.md]: The `etc` was referring to the other forms of reading.

```suggestion
All data mappings are executed during `advance` and not in `readData`, cf. the section on  [how to couple your own code](couple-your-code-overview.html).
```

File: pages/docs/couple-your-code/couple-your-code-defining-mesh-connectivity.md [hunk 2]
```diff
@@ -15,24 +15,22 @@ For volume coupling in 2D, mesh connectivity boils down to defining triangles an
...
-int setMeshEdge (int meshID, int firstVertexID, int secondVertexID);
-void setMeshTriangle (int meshID, int firstEdgeID, int secondEdgeID, int thirdEdgeID);
-void setMeshTriangleWithEdges (int meshID, int firstVertexID, int secondVertexID, int thirdVertexID);
-void setMeshQuad(int meshID, int firstEdgeID, int secondEdgeID, int thirdEdgeID, int fourthEdgeID);
-void setMeshQuadWithEdges(int meshID, int firstVertexID, int secondVertexID, int thirdVertexID, int fourthVertexID);
-void setMeshTetrahredron(int meshID, int firstVertexID, int secondVertexID, int thirdVertexID, int fourthVertexID);
+int setMeshEdge (::precice::string_view meshName, int firstVertexID, int secondVertexID);
+void setMeshTriangle (::precice::string_view meshName, int firstVertexID, int secondVertexID, int thirdVertexID);
+void setMeshQuad(::precice::string_view meshName, int firstVertexID, int secondVertexID, int thirdVertexID, int fourthVertexID);
+void setMeshTetrahedron(::precice::string_view meshName, int firstVertexID, int secondVertexID, int thirdVertexID, int fourthVertexID);
```
Comment (MakisH) [inline pages/docs/couple-your-code/couple-your-code-defining-mesh-connectivity.md]: Do we explicitly want/need the root namespace? If yes, we should clarify this somewhere. If not, let's remove it:

```suggestion
int setMeshEdge (precice::string_view meshName, int firstVertexID, int secondVertexID);
void setMeshTriangle (precice::string_view meshName, int firstVertexID, int secondVertexID, int thirdVertexID);
void setMeshQuad(precice::string_view meshName, int firstVertexID, int secondVertexID, int thirdVertexID, int fourthVertexID);
void setMeshTetrahedron(precice::string_view meshName, int firstVertexID, int secondVertexID, int thirdVertexID, int fourthVertexID);
```

File: pages/docs/couple-your-code/couple-your-code-defining-mesh-connectivity.md [hunk 3]
```diff
@@ -82,19 +80,19 @@ The order of vertices also does not matter. Triangles BAC and CAB are considered
...
 
 ```cpp
-void setMeshEdge(int meshID, int firstVertexID, int secondVertexID);
-void setMeshTriangle(int meshID, int firstVertexID, int secondVertexID, int thirdVertexID);
-void setMeshQuad(int meshID, int firstVertexID, int secondVertexID, int thirdVertexID, int fourthVertexID);
-void setMeshTetrahredron(int meshID, int firstVertexID, int secondVertexID, int thirdVertexID, int fourthVertexID);
+void setMeshEdge(::precice::string_view meshName, int firstVertexID, int secondVertexID);
+void setMeshTriangle(::precice::string_view meshName, int firstVertexID, int secondVertexID, int thirdVertexID);
+void setMeshQuad(::precice::string_view meshName, int firstVertexID, int secondVertexID, int thirdVertexID, int fourthVertexID);
+void setMeshTetrahredron(::precice::string_view meshName, int firstVertexID, int secondVertexID, int thirdVertexID, int fourthVertexID);
```
Comment (MakisH) [inline pages/docs/couple-your-code/couple-your-code-defining-mesh-connectivity.md]:
```suggestion
void setMeshEdge(precice::string_view meshName, int firstVertexID, int secondVertexID);
void setMeshTriangle(precice::string_view meshName, int firstVertexID, int secondVertexID, int thirdVertexID);
void setMeshQuad(precice::string_view meshName, int firstVertexID, int secondVertexID, int thirdVertexID, int fourthVertexID);
void setMeshTetrahredron(precice::string_view meshName, int firstVertexID, int secondVertexID, int thirdVertexID, int fourthVertexID);
```

File: pages/docs/couple-your-code/couple-your-code-defining-mesh-connectivity.md [hunk 4]
```diff
@@ -82,19 +80,19 @@ The order of vertices also does not matter. Triangles BAC and CAB are considered
...
 
 ```cpp
-void setMeshEdges(int meshID, int size, int* vertices);
-void setMeshTriangles(int meshID, int size, int* vertices);
-void setMeshQuads(int meshID, int size, int* vertices);
-void setMeshTetrahedra(int meshID, int size, int* vertices);
+void setMeshEdges(::precice::string_view meshName, ::precice::span<const VertexID> vertices);
+void setMeshTriangles(::precice::string_view meshName, ::precice::span<const VertexID> vertices);
+void setMeshQuads(::precice::string_view meshName, ::precice::span<const VertexID> vertices);
+void setMeshTetrahedra(::precice::string_view meshName, ::precice::span<const VertexID> vertices);
```
Comment (MakisH) [inline pages/docs/couple-your-code/couple-your-code-defining-mesh-connectivity.md]:
```suggestion
void setMeshEdges(precice::string_view meshName, ::precice::span<const VertexID> vertices);
void setMeshTriangles(precice::string_view meshName, ::precice::span<const VertexID> vertices);
void setMeshQuads(precice::string_view meshName, ::precice::span<const VertexID> vertices);
void setMeshTetrahedra(precice::string_view meshName, ::precice::span<const VertexID> vertices);
```

File: pages/docs/couple-your-code/couple-your-code-gradient-data.md [hunk 5]
```diff
@@ -10,42 +10,16 @@ This feature is available since version 2.4.0.
...
-void writeBlockVectorGradientData(
-    int           dataID,
-    int           size,
-    const int*    valueIndices,
-    const double* gradientValues);
+void writeGradientData(
+    ::precice::string_view          meshName,
+    ::precice::string_view          dataName,
+    ::precice::span<const VertexID> vertices,
+    ::precice::span<const double>   gradients);
```
Comment (MakisH) [inline pages/docs/couple-your-code/couple-your-code-gradient-data.md]:
```suggestion
    precice::string_view          meshName,
    precice::string_view          dataName,
    precice::span<const VertexID> vertices,
    precice::span<const double>   gradients);
```

File: pages/docs/couple-your-code/couple-your-code-mesh-and-data-access.md [hunk 6]
```diff
@@ -10,29 +10,31 @@ For coupling, we need coupling meshes. Let's see how we can tell preCICE about o
...
-int setMeshVertex (int meshID, const double* position);
-void setMeshVertices (int meshID, int size, double* positions, int* ids);
+int setMeshVertex(
+    ::precice::string_view        meshName,
+    ::precice::span<const double> position);
+
+void setMeshVertices(
+    ::precice::string_view        meshName,
+    ::precice::span<const double> positions,
+    ::precice::span<VertexID>     ids);
```
Comment (MakisH) [inline pages/docs/couple-your-code/couple-your-code-mesh-and-data-access.md]:
```suggestion
int setMeshVertex(
    ::precice::string_view        meshName,
    ::precice::span<const double> position);

void setMeshVertices(
    precice::string_view        meshName,
    precice::span<const double> positions,
    precice::span<VertexID>     ids);
```

File: pages/docs/couple-your-code/couple-your-code-mesh-and-data-access.md [hunk 7]
```diff
@@ -10,29 +10,31 @@ For coupling, we need coupling meshes. Let's see how we can tell preCICE about o
...
 
 ```cpp
-int getDataID (const std::string& dataName, int meshID);
-void writeVectorData (int dataID, int vertexID, const double* value);
-void writeBlockVectorData (int dataID, int size, int* vertexIDs, double* values);
+void Participant::writeData(
+    ::precice::string_view          meshName,
+    ::precice::string_view          dataName,
+    ::precice::span<const VertexID> vertices,
+    ::precice::span<const double>   values)
```
Comment (MakisH) [inline pages/docs/couple-your-code/couple-your-code-mesh-and-data-access.md]:
```suggestion
void Participant::writeData(
    precice::string_view          meshName,
    precice::string_view          dataName,
    precice::span<const VertexID> vertices,
    precice::span<const double>   values)
```

File: pages/docs/couple-your-code/couple-your-code-mesh-and-data-access.md [hunk 8]
```diff
@@ -10,29 +10,31 @@ For coupling, we need coupling meshes. Let's see how we can tell preCICE about o
...
 ```
 
-* `getDataID` returns the data ID for a coupling data field (e.g. "Displacements", "Forces", etc).
-* `writeVectorData` writes vector-valued data to the coupling data structure.
-* `writeBlockVectorData` writes multiple vector data at once, again for performance reasons.
-
-Similarly, there are methods for reading coupling data: `readVectorData` and `readBlockVectorData`. Furthermore,
-preCICE distinguishes between scalar-valued and vector-valued data. For scalar data, similar methods exist, for example `writeScalarData`.
+<!-- Also point to section where `relativeReadTime` is explained? We will probably solve this in https://github.com/precice/precice.github.io/pull/257 -->
```
Comment (MakisH) [inline pages/docs/couple-your-code/couple-your-code-mesh-and-data-access.md]:
```suggestion
<!-- TODO Also point to section where `relativeReadTime` is explained? We will probably solve this in https://github.com/precice/precice.github.io/pull/257 -->
```

(yes, we probably should)

File: pages/docs/couple-your-code/couple-your-code-waveform.md [hunk 9]
```diff
@@ -33,22 +33,24 @@ Linear interpolation between coupling boundary conditions of the previous and th
...
-void readBlockVectorData(int dataID, int size, const int* valueIndices, double* values) const;
-
-// experimental API for waveform iteration
-void readBlockVectorData(int dataID, int size, const int* valueIndices, double relativeReadTime, double* values) const;
+void Participant::readData(
+    ::precice::string_view          meshName,
+    ::precice::string_view          dataName,
+    ::precice::span<const VertexID> vertices,
+    double                          relativeReadTime,
+    ::precice::span<double>         values) const
```
Comment (MakisH) [inline pages/docs/couple-your-code/couple-your-code-waveform.md]:
```suggestion
    precice::string_view          meshName,
    precice::string_view          dataName,
    precice::span<const VertexID> vertices,
    double                          relativeReadTime,
    precice::span<double>         values) const
```

**Review by MakisH — COMMENTED** [review]
Some first comments, there are anyway still some `TODO`s.
Thanks a lot for the effort of updating the API! This is very important and helpful.

**Comment by MakisH** [comment]
Do you want to handle the `TODO` parts in this PR, or in a separate one?

**Comment by BenjaminRodenberg** [comment]
> Do you want to handle the `TODO` parts in this PR, or in a separate one?

I think we should merge this one soon. If somebody else want's to handle the TODO parts while reviewing, I'm fine. If not, we can do it later.

**Review by MakisH — APPROVED** [review]
Then, feel free to open an issue and merge this.
FYI, I have not cross-checked with the documentation whether there is more than find&replace needed in the points you have touched.

**Pull request merged**